### PR TITLE
Implement "distroless" containers

### PR DESCRIPTION
**Pull request by MateuszMalisz**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

This PR enables users to create a "distroless" container from Mariner. The distroless container does not contain the whole OS (e.g. kernel), but rather a minimal set of packages required for it to work as a container. This minimizes the attack surface and space requirements.

###### Change Log  

- Remove requirements for bash for some packages with `%post`. Bash will be required by package managers instead.
- Add busybox and uclibc packages
- Remove RPM database from the image if it does not contain a package manager.
- Implement chage functionality to not depend on the binary.
- Fix cleanup of the additional files in the chroot.

###### Does this affect the toolchain?  

**YES**

###### Test Methodology

- Pipeline build id: 41101

### CODE REVIEW

**Review by schmittjoseph — CHANGES_REQUESTED** [review]

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 1]
```diff
@@ -922,14 +942,110 @@ func createUserWithPassword(installChroot *safechroot.Chroot, user configuration
...
+		done := false
+		// Entries in shadow are separated by colon and start with a username
+		// Finding one that starts like that means we've found our entry
+		if strings.HasPrefix(entry, usernameWithColon) {
+			// Each line in shadow contains 9 fields separated by colon ("") in the following order:
+			// login name, encrypted password, date of last password change,
+			// minimum password age, maximum password age, password warning period,
+			// password inactivity period, account expiration date, reserved field for future use
+			const (
+				loginNameField         = 0
```
[the remark this answers is not in the document]
Comment (MateuszMalisz) [reply toolkit/tools/imagegen/installutils/installutils.go:989]: I thought about using `iota`, but since this is not an enumeration, the fields have the numbers assigned to them and they are immutable - I think it's better to be explicit about position of each field.

File: SPECS/glibc/glibc.spec [hunk 2]
```diff
@@ -1,6 +1,8 @@
 %global security_hardening nonow
 %define glibc_target_cpu %{_build}
 %define debug_package %{nil}
+# Don't depend on bash by default - distroless
```
Comment (MateuszMalisz) [inline SPECS/glibc/glibc.spec:4]: remove the `- distroless`

File: SPECS/distroless-packages/distroless-packages.spec [hunk 3]
```diff
@@ -0,0 +1,51 @@
+Name:           distroless-packages
+Summary:        Metapackage with core sets of packages for distroless containers
+Version:        0.1
+Release:        1%{?dist}
```
Comment (MateuszMalisz) [inline SPECS/distroless-packages/distroless-packages.spec:4]: Fix release

File: SPECS/openssl/openssl.spec [hunk 4]
```diff
@@ -1,3 +1,6 @@
+# Don't depend on bash by default - distroless
```
Comment (MateuszMalisz) [inline SPECS/openssl/openssl.spec:1]: Remove `- distroless`

File: toolkit/resources/manifests/package/toolchain_x86_64.txt [hunk 5]
```diff
@@ -17,12 +17,12 @@ bzip2-1.0.6-15.cm1.x86_64.rpm
...
-ca-certificates-legacy-20200720-9.cm1.noarch.rpm
-ca-certificates-microsoft-20200720-9.cm1.noarch.rpm
-ca-certificates-shared-20200720-9.cm1.noarch.rpm
-ca-certificates-tools-20200720-9.cm1.noarch.rpm
+ca-certificates-20200720-10.cm1.noarch.rpm
+ca-certificates-base-20200720-10.cm1.noarch.rpm
+ca-certificates-legacy-20200720-10.cm1.noarch.rpm
+ca-certificates-microsoft-20200720-10.cm1.noarch.rpm
+ca-certificates-shared-20200720-10.cm1.noarch.rpm
+ca-certificates-tools-20200720-10.cm1.noarch.rpm
```
Comment (schmittjoseph) [inline toolkit/resources/manifests/package/toolchain_x86_64.txt:25]: Add `ca-certificates-static` to this list (and the aarch64 manifest)

File: toolkit/resources/manifests/package/pkggen_core_x86_64.txt [hunk 6]
```diff
@@ -161,6 +161,6 @@ libffi-3.2.1-12.cm1.x86_64.rpm
 libtasn1-4.14-2.cm1.x86_64.rpm
 p11-kit-0.23.16.1-2.cm1.x86_64.rpm
 p11-kit-trust-0.23.16.1-2.cm1.x86_64.rpm
-ca-certificates-shared-20200720-9.cm1.noarch.rpm
-ca-certificates-tools-20200720-9.cm1.noarch.rpm
-ca-certificates-base-20200720-9.cm1.noarch.rpm
+ca-certificates-shared-20200720-10.cm1.noarch.rpm
+ca-certificates-tools-20200720-10.cm1.noarch.rpm
+ca-certificates-base-20200720-10.cm1.noarch.rpm
```
Comment (schmittjoseph) [inline toolkit/resources/manifests/package/pkggen_core_x86_64.txt:166]: If you don't want `ca-certificates-static` in the worker chroot, you'll also need to update `toolkit/resources/manifests/package/update_manifests.sh`.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 7]
```diff
@@ -900,8 +913,16 @@ func createUserWithPassword(installChroot *safechroot.Chroot, user configuration
 			logger.Log.Warnf("Ignoring UID for (%s) user, using default", rootUser)
 		}
 
-		// Update shadow file
-		err = updateUserPassword(installChroot.RootDir(), user.Name, hashedPassword)
+		if _, err2 := os.Stat(shadowFile); err2 != nil {
```
Comment (schmittjoseph) [inline toolkit/tools/imagegen/installutils/installutils.go:916]: Can switch to `file.PathExists` here as well.

File: toolkit/tools/imagegen/installutils/installutils.go [hunk 8]
```diff
@@ -922,14 +943,109 @@ func createUserWithPassword(installChroot *safechroot.Chroot, user configuration
 
 	// Update password expiration
 	if user.PasswordExpiresDays != 0 {
+		// Ignore updating if there is no shadow file to update
+		if _, err2 := os.Stat(shadowFile); err2 != nil {
```
Comment (schmittjoseph) [inline toolkit/tools/imagegen/installutils/installutils.go:947]: Can switch to `file.PathExists` here as well.

**Review by schmittjoseph — CHANGES_REQUESTED** [review]

File: toolkit/resources/manifests/package/toolchain_x86_64.txt (hunk 5 above)
Comment (MateuszMalisz) [reply toolkit/resources/manifests/package/toolchain_x86_64.txt:25]: Good catch!

**Review by schmittjoseph — APPROVED** [review]

File: SPECS/busybox/busybox.spec [hunk 9]
```diff
@@ -0,0 +1,603 @@
+Summary: Statically linked binary providing simplified versions of system commands
+Name:    busybox
+Version: 1.31.1
+Release: 3%{?dist}
+License: GPLv2
+URL:     http://www.busybox.net
```
Comment (PawelWMS) [inline SPECS/busybox/busybox.spec:6]: Please use HTTPS.

File: SPECS/busybox/busybox.spec [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+
+Source:  http://www.busybox.net/downloads/%{name}-%{version}.tar.bz2
```
Comment (PawelWMS) [inline SPECS/busybox/busybox.spec:8]: Please use `%{url}` to minimize repretition.

File: SPECS/busybox/busybox.spec [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+%install
+mkdir -p $RPM_BUILD_ROOT/sbin
+install -m 755 busybox.static $RPM_BUILD_ROOT/sbin/busybox
+install -m 755 busybox.petitboot $RPM_BUILD_ROOT/sbin/busybox.petitboot
+mkdir -p $RPM_BUILD_ROOT/%{_mandir}/man1
+install -m 644 docs/busybox.static.1 $RPM_BUILD_ROOT/%{_mandir}/man1/busybox.1
+install -m 644 docs/busybox.petitboot.1 $RPM_BUILD_ROOT/%{_mandir}/man1/busybox.petitboot.1
+
+%files
+%doc LICENSE README
```
Comment (PawelWMS) [inline SPECS/busybox/busybox.spec:116]: Please use the `%license` macro.

File: SPECS/busybox/busybox.spec [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+%{_mandir}/man1/busybox.1.gz
+
+%files petitboot
+%doc LICENSE README
+/sbin/busybox.petitboot
+%{_mandir}/man1/busybox.petitboot.1.gz
+
+%changelog
+* Thu Oct 15 2020 Mateusz Malisz <mamalisz@microsoft.com> - 1.31.1-3
+- Initial CBL-Mariner import from Fedora 32 (license: MIT)
```
Comment (PawelWMS) [inline SPECS/busybox/busybox.spec:127]: Please add "License verified".

File: SPECS/ca-certificates/ca-certificates.spec [hunk 13]
```diff
@@ -189,6 +189,14 @@ Requires: %{name}-shared = %{version}-%{release}
 Provides a legacy version of ca-bundle.crt in the format of "[hash].0 -> [hash].pem"
 pairs under /etc/pki/tls/certs.
 
+%package static
+Summary:  ca-certs package with minimized runtime requirements
+Group:    System Environment/Security
+Requires: %{name}-shared = %{version}-%{release}
+
+%description static
+Provides ca-certificates while minimizing the runtime dependencies. This package does not contain the post section.
```
Comment (PawelWMS) [inline SPECS/ca-certificates/ca-certificates.spec:198]: The `%post` section is what looks at the sources and puts the certificates in the right places for them to be usable to different system components. I'm not sure what's the purpose of this particular subpackage.

File: SPECS/ca-certificates/ca-certificates.spec [hunk 14]
```diff
@@ -363,6 +371,21 @@ rm -f %{pkidir}/tls/certs/*.{0,pem}
...
+%{_datadir}/pki/ca-trust-legacy/%{legacy_default_base_bundle}
+%{_datadir}/pki/ca-trust-legacy/%{legacy_disable_base_bundle}
+# main pkg files
+%{_datadir}/pki/ca-trust-source/%{p11_format_mozilla_bundle}
+%{_datadir}/pki/ca-trust-legacy/%{legacy_default_mozilla_bundle}
+%{_datadir}/pki/ca-trust-legacy/%{legacy_disable_mozilla_bundle}
+# microsoft files
+%{_datadir}/pki/ca-trust-source/%{p11_format_microsoft_bundle}
+%{_datadir}/pki/ca-trust-legacy/%{legacy_default_microsoft_bundle}
+%{_datadir}/pki/ca-trust-legacy/%{legacy_disable_microsoft_bundle}
```
Comment (PawelWMS) [inline SPECS/ca-certificates/ca-certificates.spec:386]: We MUST NOT install all sources for different certificate packs with one subpackage. They are kept separate by design. Usually, a basic Mariner image should only need the ones from the `*-base` subpackage. The expectation is to trust only the minimial set of entities. Expending that trust is up to the user and we cannot implicitly trust anything else for them.

File: SPECS/distroless-packages/distroless-packages.spec [hunk 15]
```diff
@@ -0,0 +1,51 @@
+Name:           distroless-packages
+Summary:        Metapackage with core sets of packages for distroless containers
+Version:        0.1
+Release:        2%{?dist}
+License:        MIT
+Group:          System Environment/Base
+URL:            http://aka.ms/cbl-mariner
```
Comment (PawelWMS) [inline SPECS/distroless-packages/distroless-packages.spec:7]: Please use HTTPS and ideally the final URL, not the shortcut, so the user knows where they're going.

File: SPECS/distroless-packages/distroless-packages.spec [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+%files debug
+
+%changelog
+* Thu Oct 15 2020 Mateusz Malisz <mamalisz@microsoft.com> - 0.1-2
+- Extend the set of requirements for the base image
+- Add debug package with busybox
+
+* Tue Sep 01 2020 Jon Slobodzian <joslobo@microsoft.com> - 0.1-1
+- Initial Mariner Version
```
Comment (PawelWMS) [inline SPECS/distroless-packages/distroless-packages.spec:51]: Outdated, please use `Original version for CBL-Mariner.`.

**Comment by PawelWMS** [comment]
> 

Please update changelog.

---
Refers to: SPECS/glibc/glibc.spec:303 in 20d1d14. [](commit_id = 20d1d14d669d5c0839d6e85e2861b9854273c4d3, deletion_comment = False)

File: SPECS/openssl/openssl.spec [hunk 17]
```diff
@@ -38,7 +40,6 @@ Group: System Environment/Libraries
 OpenSSL is a toolkit for supporting cryptography. The openssl-libs
 package contains the libraries that are used by various applications which
 support cryptographic algorithms and protocols.
-Requires: openssl = %{version}-%{release}
```
Comment (PawelWMS) [inline SPECS/openssl/openssl.spec:41]: Double-checking: are you sure the libs don't need the default package?

**Comment by PawelWMS** [comment]
> * Tue Jul 28 2020 Pawel Winogrodzki <pawelwi@microsoft.com> 1.1.1g-6

Missing changelog entry.

---
Refers to: SPECS/openssl/openssl.spec:261 in 20d1d14. [](commit_id = 20d1d14d669d5c0839d6e85e2861b9854273c4d3, deletion_comment = False)

File: SPECS/uclibc/uclibc.spec [hunk 18]
```diff
@@ -0,0 +1,198 @@
+%global uclibc_name uClibc
+
+Name:    uclibc
+Version: 0.9.33.2
+Release: 22%{?dist}
+Summary: C library for embedded Linux
+
+License: LGPLv2
+URL:     http://www.uclibc.org/
```
Comment (PawelWMS) [inline SPECS/uclibc/uclibc.spec:9]: Please use HTTPS.

File: SPECS/uclibc/uclibc.spec [hunk 19]
```diff
@@ -0,0 +1,198 @@
+%global uclibc_name uClibc
```
Comment (PawelWMS) [inline SPECS/uclibc/uclibc.spec:1]: Nit: seems unnecessary. The source URL is case-insensitive, so we can just directly use this for the `%setup` macro.

File: SPECS/uclibc/uclibc.spec [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+%files devel
+%doc README docs/Glibc_vs_uClibc_Differences.txt docs/threads.txt docs/uClibc_vs_SuSv3.txt
+%doc TODO DEDICATION.mjn3 MAINTAINERS
+%doc docs/PORTING COPYING.LIB
+%{_includedir}/uClibc
+%{_libdir}/uClibc
+
+%changelog
+* Thu Oct 15 2020 Mateusz Malisz <mamalisz@microsoft.com> - 0.9.33.2-22
+- Initial CBL-Mariner import from Fedora 32 (license: MIT)
```
Comment (PawelWMS) [inline SPECS/uclibc/uclibc.spec:90]: Please add "License verified". Of course after validating the license firt.

File: SPECS/uclibc/uclibc.spec [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+mkdir -p $RPM_BUILD_ROOT/%{_libdir}/uClibc
+mv  $RPM_BUILD_ROOT/lib/*  $RPM_BUILD_ROOT/%{_libdir}/uClibc/
+rm -rf  $RPM_BUILD_ROOT/lib/
+
+# move the header files to /usr subdirectory
+mkdir -p $RPM_BUILD_ROOT/%{_includedir}/uClibc
+mv  $RPM_BUILD_ROOT/include/*  $RPM_BUILD_ROOT/%{_includedir}/uClibc
+rm -rf  $RPM_BUILD_ROOT/include/
+
+%files devel
```
Comment (PawelWMS) [inline SPECS/uclibc/uclibc.spec:81]: Missing `%license` macro.

File: SPECS/ca-certificates/ca-certificates.spec (hunk 14 above)
Comment (PawelWMS) [inline SPECS/ca-certificates/ca-certificates.spec:386]: Why is this change required? I believe you've mentioned it to me a few weeks ago, but I no longer recall. The `*-shared` subpackage installed all the directories and files required by subpackages actually installing the certificates. This seems now broken.
